cpp/daal/src/data_management/finiteness_checker_cpu.cpp

```diff
@@ -0,0 +1,917 @@
+    else
+        *((uint64_t *)(&inf)) = doubleExpMask;
+    return inf;
+}
+
+    #if defined(__AVX2__) and !defined(__AVX512F__)
+
+template <typename DataType>
+DataType sumWithAVX2(size_t n, const DataType * dataPtr)
+{
+    const size_t nPerInstr = 32 / sizeof(DataType);
+    DataType sum;
+    if (sizeof(DataType) == 4)
+    {
+        __m256 sums     = _mm256_set1_ps(0);
+        __m256 * ptr256 = (__m256 *)dataPtr;
+        for (size_t i = 0; i < n / nPerInstr; i++) sums = _mm256_add_ps(sums, ptr256[i]);
+
+        // AVX2 doesn't have reduce_add_ps, so finer-grained used of intrinsics necessary
+        __m128 s0    = _mm256_castps256_ps128(sums);
+        __m128 s1    = _mm256_extractf128_ps(sums, 1);
+        s0           = _mm_add_ps(s0, s1);
+        __m128 stemp = _mm_movehl_ps(s0, s0);
+        __m128 sD    = _mm_add_ps(s0, stemp);
+        __m128 hi    = _mm_shuffle_ps(sD, sD, 0x1);
+
+        sum = _mm_cvtss_f32(_mm_add_ss(sD, hi));
+    }
+    else
+    {
+        __m256d sums     = _mm256_set1_pd(0);
+        __m256d * ptr256 = (__m256d *)dataPtr;
+        for (size_t i = 0; i < n / nPerInstr; i++) sums = _mm256_add_pd(sums, ptr256[i]);
+
+        // AVX2 doesn't have reduce_add_pd, so finer-grained used of intrinsics necessary
+        __m128d s0    = _mm256_castpd256_pd128(sums);
+        __m128d s1    = _mm256_extractf128_pd(sums, 1);
+        s0            = _mm_add_pd(s0, s1);
+        __m128d stemp = _mm_unpackhi_pd(s0, s0);
+
+        sum = _mm_cvtsd_f64(_mm_add_sd(s0, stemp));
+    }
+    for (size_t i = (n / nPerInstr) * nPerInstr; i < n; ++i) sum += dataPtr[i];
+
+    return sum;
+}
+
+template <typename DataType>
+DataType computeSumAVX2Impl(size_t nDataPtrs, size_t nElementsPerPtr, const DataType ** dataPtrs)
+{
+    size_t nBlocksPerPtr = nElementsPerPtr / BLOCK_SIZE;
+    if (nBlocksPerPtr == 0) nBlocksPerPtr = 1;
+    size_t nElements    = nDataPtrs * nElementsPerPtr;
+    bool inParallel     = !(nElements < THREADING_BORDER);
+    size_t nPerBlock    = nElementsPerPtr / nBlocksPerPtr;
+    size_t nSurplus     = nElementsPerPtr % nBlocksPerPtr;
+    size_t nTotalBlocks = nBlocksPerPtr * nDataPtrs;
+
+    daal::services::internal::TArray<DataType, avx2> partialSumsArr(nTotalBlocks);
+    DataType * pSums = partialSumsArr.get();
+    if (!pSums) return getInf<DataType>();
+    for (size_t iBlock = 0; iBlock < nTotalBlocks; ++iBlock) pSums[iBlock] = 0;
+
+    daal::conditional_threader_for(inParallel, nTotalBlocks, [&](size_t iBlock) {
+        size_t ptrIdx        = iBlock / nBlocksPerPtr;
+        size_t blockIdxInPtr = iBlock - nBlocksPerPtr * ptrIdx;
+        size_t start         = blockIdxInPtr * nPerBlock;
+        size_t end           = blockIdxInPtr == nBlocksPerPtr - 1 ? start + nPerBlock + nSurplus : start + nPerBlock;
+
+        pSums[iBlock] = sumWithAVX2<DataType>(end - start, dataPtrs[ptrIdx] + start);
+    });
+
+    return sumWithAVX2<DataType>(nTotalBlocks, pSums);
+}
```

Comment:
This code is almost 100% copy-paste from AVX512 code.
I think it can be unified if sumWithXXX and computeSumXXXImpl functions would have a `cpu` template parameter.